feat(lighter): new exchange

sc0Vu opened the pull request:
```BASH
ph lighter createOrder ETH/USDC:USDC limit sell 0.005 3600 '{"timeInForce":"gtt"}'
ph lighter createOrder ETH/USDC:USDC limit sell 0.005 3600 '{"timeInForce":"ioc"}'
ph lighter createOrder ETH/USDC:USDC limit sell 0.005 3600 '{"timeInForce":"gtt","stopLossPrice":3100}'
ph lighter createOrder ETH/USDC:USDC limit sell 0.005 3600 '{"timeInForce":"gtt","stopLossPrice":2900}'
ph lighter createOrder ETH/USDC:USDC limit buy 0.005 2000 '{"timeInForce":"gtt","stopLossPrice":2900}'
ph lighter createOrder ETH/USDC:USDC limit buy 0.005 2000 '{"timeInForce":"gtt","stopLossPrice":3400}'
ph lighter createOrder ETH/USDC:USDC limit buy 0.005 2000 '{"timeInForce":"gtt","takeProfitPrice":3400}'
ph lighter createOrder ETH/USDC:USDC limit buy 0.005 2000 '{"timeInForce":"gtt","stopLossPrice":3400,"clientOrderId":123456789}'

ph lighter createOrder ETH/USDC:USDC limit sell 0.005 3500 '{"timeInForce":"ioc","takeProfit":{"type":"limit","triggerPrice":2800,"price":2700},"stopLoss":{"type":"limit","triggerPrice":3700,"price":3800}}'
ph lighter createOrder ETH/USDC:USDC limit sell 0.005 3500 '{"timeInForce":"gtt","takeProfit":{"type":"limit","triggerPrice":2800,"price":2700},"stopLoss":{"type":"limit","triggerPrice":3700,"price":3800}}'

ph lighter fetchOpenOrders ETH/USDC:USDC
ph lighter cancelOrder 562948951840191 ETH/USDC:USDC
ph lighter cancelAllOrders
ph lighter cancelAllOrdersAfter 300000



ph lighter fetchStatus
ph lighter fetchTime
ph lighter fetchPositions
ph lighter fetchAccounts
ph lighter fetchBalance '{"apiKeyIndex":4,"accountIndex":668800}'
ph lighter fetchClosedOrders ETH/USDC:USDC
ph lighter createOrder ETH/USDC:USDC limit buy 0.005 2000 '{"timeInForce":"gtt","stopLossPrice":3400,"clientOrderId":123456789}'
ph lighter createOrder ETH/USDC:USDC limit buy 0.005 2000 '{"timeInForce":"gtt","takeProfitPrice":3400}'
ph lighter fetchOpenOrders ETH/USDC:USDC
ph lighter fetchDeposits USDC 0 1 '{"address":"0x754e6a9E4BC74D119263FD249fAc5d876D0D10eE"}'
ph lighter withdraw USDC 6 "0x754e6a9E4BC74D119263FD249fAc5d876D0D10eE"
ph lighter setLeverage 1 ETH/USDC:USDC '{"marginMode":"isolated"}'
ph lighter setLeverage 1 ETH/USDC:USDC '{"marginMode":"cross"}'
ph lighter setMarginMode cross ETH/USDC:USDC '{"leverage":1}'


cs lighter createOrder ETH/USDC:USDC limit sell 0.005 3900 '{"timeInForce":"gtt","apiKeyIndex":4,"accountIndex":668800}'
cs lighter createOrder ETH/USDC:USDC limit sell 0.005 3900 '{"timeInForce":"gtt","apiKeyIndex":4,"accountIndex":668800,"takeProfit":{"type":"limit","triggerPrice":2800,"price":2700},"stopLoss":{"type":"limit","triggerPrice":3700,"price":3800}}'
cs lighter fetchOpenOrders ETH/USDC:USDC 0 1 '{"apiKeyIndex":4,"accountIndex":668800}'
cs lighter editOrder 562948923372019 ETH/USDC:USDC limit sell 0.005 4000 '{"timeInForce":"gtt","apiKeyIndex":4,"accountIndex":668800}'
cs lighter cancelOrder 281476032796048 ETH/USDC:USDC '{"apiKeyIndex":4,"accountIndex":668800}'
cs lighter cancelAllOrders ETH/USDC:USDC '{"apiKeyIndex":4,"accountIndex":668800}'
cs lighter transfer USDC 1 perp spot '{"memo":"0x000000000000000000000000006465","apiKeyIndex":4,"accountIndex":668800}'
cs lighter setLeverage 1 ETH/USDC:USDC '{"marginMode":"cross","apiKeyIndex":4,"accountIndex":668800}'
cs lighter setMarginMode cross ETH/USDC:USDC '{"leverage":1,"apiKeyIndex":4,"accountIndex":668800}'
cs lighter addMargin ETH/USDC:USDC 1 '{"apiKeyIndex":4,"accountIndex":668800}'
cs lighter createSubAccount '{"libraryPath":"/Users/laiguancheng/Desktop/projects/opensource/lighter-go/build/lighter-signer-darwin-x86.dylib","apiKeyIndex":4,"accountIndex":668800}'
```
<img width="597" height="192" alt="dark_logo_full" src="https://github.com/user-attachments/assets/b227f3e2-60d9-4163-bedf-807adfd57d1e" />

sc0Vu committed:
feat(lighter): new exchange

sc0Vu committed:
feat(lighter): add fetchStatus & fetchTime

sc0Vu committed:
feat(lighter): add fetchMarkets

sc0Vu committed:
feat(lighter): add fetchOrderBook

sc0Vu committed:
feat(lighter): add fetchTicker

sc0Vu committed:
feat(lighter): add fetchTickers

sc0Vu committed:
feat(lighter): add static-tests

sc0Vu committed:
feat(lighter): add fetchOHLCV

sc0Vu committed:
feat(lighter): add fetchFundingRates

sc0Vu committed:
feat(lighter): add watchOrderBook

sc0Vu committed:
feat(lighter): add unWatchOrderBook

sc0Vu committed:
feat(lighter): add watchTicker & unWatchTicker

sc0Vu committed:
feat(lighter): add watchTickers & unWatchTickers

sc0Vu committed:
feat(lighter): add watchMarkPrice & watchMarkPrices

sc0Vu committed:
feat(lighter): add handlePing

sc0Vu committed:
feat(lighter): add watchTrades

sc0Vu committed:
feat(lighter): add watchLiquidations

sc0Vu committed:
Merge branch 'master' into add-lighter

sc0Vu committed:
Merge branch 'master' into add-lighter

sc0Vu committed:
lighter: add python package

sc0Vu committed:
feat(lighter): add private create auth

sc0Vu committed:
feat(lighter): update

sc0Vu committed:
feat(lighter): add fetchBalance

sc0Vu committed:
feat(lighter): add fetchPositions

sc0Vu committed:
lighter: update go

sc0Vu committed:
Support multipart in async python

sc0Vu committed:
Support multipart in python

sc0Vu committed:
feat(lighter): update

sc0Vu committed:
feat(lighter): add fetchPosition

sc0Vu committed:
feat(lighter): add fetchClosedOrders

sc0Vu committed:
Add lighter signer in php

sc0Vu committed:
lighter: fetch nonce in ccxt

sc0Vu committed:
go: add multipart

sc0Vu committed:
feat(lighter): add fetchOpenOrders

sc0Vu committed:
feat(lighter): add fetchCurrencies

sc0Vu committed:
feat(lighter): add fetchTransfers

sc0Vu committed:
feat(lighter): add fetchDeposits & fetchWithdrawals

sc0Vu committed:
feat(lighter): update fetchBalance

sc0Vu committed:
feat(lighter): add fetchMyTrades

sc0Vu committed:
feat(lighter): add transfer

sc0Vu committed:
feat(lighter): add fetchAccounts

sc0Vu committed:
feat(lighter): update exchange

sc0Vu committed:
feat(lighter): add setLeverage

sc0Vu committed:
lighter: update nonce

sc0Vu committed:
lighter: add pow

sc0Vu committed:
feat(lighter): fix transfer

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: add multipart/form-data in php sync

sc0Vu committed:
lighter: add multipart/form-data in php async

sc0Vu committed:
lighter: update createOrder

sc0Vu committed:
lighter: add grouped orders in python

sc0Vu committed:
feat(lighter): add setMarginMode

sc0Vu committed:
feat(lighter): add cancelOrder

sc0Vu committed:
feat(lighter): add cancelAllOrders

sc0Vu committed:
fix(lighter): fix cancelOrder

sc0Vu committed:
feat(lighter): add setMargin

sc0Vu committed:
feat(lighter): add withdraw

sc0Vu committed:
feat(lighter): update Exchange.php

sc0Vu committed:
feat(lighter): add editOrder

sc0Vu committed:
feat(lighter): add editOrder php

sc0Vu committed:
feat(lighter): add createSubAccount

sc0Vu committed:
feat(lighter): update exchange_eth.go

sc0Vu committed:
feat(lighter): go add createAuth, add SafeNum

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update go mod

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update doc

sc0Vu committed:
lighter: cancelAllOrdersAfter

sc0Vu committed:
lighter: update cancelAllOrders

sc0Vu committed:
lighter: update clientOrderId

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: add c#

sc0Vu committed:
feat(lighter): update exchange.go

sc0Vu committed:
lighter: c# create single order

sc0Vu committed:
Merge branch 'add-lighter' of github.com:sc0Vu/ccxt into add-lighter

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update grouped order in c#

sc0Vu committed:
feat(lighter): fix golang bug

sc0Vu committed:
lighter: update exchange ts

sc0Vu committed:
lighter: update

sc0Vu committed:
feat(lighter): add go lighterSignCreateGroupedOrders

carlosmiei commented on ts/src/lighter.ts:
```diff
@@ -0,0 +1,2845 @@
+    async fetchBalance (params = {}): Promise<Balances> {
+        await this.loadMarkets ();
+        let accountIndex = undefined;
+        [ accountIndex, params ] = this.handleOptionAndParams2 (params, 'fetchBalance', 'accountIndex', 'account_index');
```
what if accountIdex is undefined?

carlosmiei commented on ts/src/lighter.ts:
```diff
@@ -0,0 +1,2845 @@
+            const perpUSDC = this.safeString (account, 'collateral');
+            const perpBalance = this.safeDict (result, 'USDC(PERP)', this.account ());
+            perpBalance['total'] = Precise.stringAdd (perpBalance['total'], perpUSDC);
+            result['USDC(PERP)'] = perpBalance;
```
hum `USDC(PERP)` is not a unified code

sc0Vu committed:
lighter: update lighterCreateAuthToken c#

sc0Vu committed:
lighter: update lighterEditOrder c#

sc0Vu committed:
lighter: update lighterCancelOrder c#

sc0Vu committed:
lighter: update lighterSignCancelAllOrders c#

sc0Vu committed:
lighter: update lighterSignWithdraw c#

sc0Vu committed:
lighter: update lighterSignTransfer c#

sc0Vu committed:
lighter: update lighterSignUpdateLeverage c#

sc0Vu committed:
lighter: update lighterSignUpdateMargin c#

sc0Vu committed:
lighter: update lighterSignSubAccount c#

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
Merge branch 'master' into add-lighter

sc0Vu committed:
Merge branch 'master' into add-lighter

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

carlosmiei commented on ts/src/lighter.ts:
```diff
@@ -0,0 +1,2881 @@
+     * @returns {object} an [order structure]{@link https://docs.ccxt.com/?id=order-structure}
+     */
+    async createOrder (symbol: string, type: OrderType, side: OrderSide, amount: number, price: Num = undefined, params = {}) {
+        await this.loadMarkets ();
```
@sc0Vu it's using these values but then I get the error shown above

<img width="1036" height="1207" alt="Image" src="https://github.com/user-attachments/assets/4b603dae-318e-4538-8037-1d31913234e5" />

sc0Vu replied:
@carlosmiei Those value are required to call their api. You can set those value in options or params.

```
"options": {
            "libraryPath": "path_to_lighter/lighter-signer-xxx.xxx",
            "apiKeyIndex": xxx, // can find it on web ui
            "accountIndex": xxxxxx // can find it when sign on web
        }
    },
```

0 and 255 is the default value from their example. Hdyt we throw error if the value is absent?

carlosmiei commented on ts/src/lighter.ts:
```diff
@@ -0,0 +1,2881 @@
+                'apiKey': false,
+                'secret': false,
+                'walletAddress': false,
+                'privateKey': true,
```
@sc0Vu this does nothing if we don't call `checkRequiredCredentials`, we need to call it

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: update

carlosmiei committed:
improve error message

carlosmiei committed:
add logo

carlosmiei committed:
add isLighterLibraryPathRequired

carlosmiei committed:
chore: init lighter wasm loading JS

carlosmiei committed:
check error and add createOrder to ts

carlosmiei committed:
some mroe error checks

carlosmiei committed:
implement types in JS

carlosmiei committed:
read path to wasm_exec.js

carlosmiei committed:
error check and make load_lighter_library async

sc0Vu committed:
Merge pull request #2 from carlosmiei/add-lighter-wasm

lighter: load WASM version in JS

sc0Vu committed:
lighter: update

sc0Vu committed:
lighter: support form-data

carlosmiei committed:
several fixes

carlosmiei committed:
adding placeholder

carlosmiei committed:
allow nonce in options

carlosmiei committed:
try add async wrapper

carlosmiei committed:
fix linting

carlosmiei committed:
skip loadLibrary on browser

carlosmiei committed:
make error message clear

carlosmiei committed:
add per language error

carlosmiei committed:
fix c# build

carlosmiei committed:
some defaults and fixes

carlosmiei committed:
add image to FAQ

carlosmiei committed:
add lighter binaries to test

carlosmiei committed:
require price

carlosmiei committed:
fix token key

carlosmiei committed:
fix auth calls

carlosmiei committed:
loadAccount inside fetchClosedOrders

carlosmiei committed:
load account in some more methods

carlosmiei committed:
fixes and tests

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into add-lighter

carlosmiei committed:
fix JS browser bundle

carlosmiei committed:
fix python linting and redef

carlosmiei committed:
fix import

carlosmiei committed:
add debug logic

carlosmiei committed:
fix filesystem issue

carlosmiei committed:
fix linting

carlosmiei committed:
update required binary

carlosmiei committed:
add info

carlosmiei committed:
update file info

carlosmiei committed:
add debug step

carlosmiei committed:
check checksum

carlosmiei committed:
add lfs version

carlosmiei committed:
add more debugging logs

carlosmiei committed:
renormalize binaries with gitattributes

carlosmiei committed:
readd binaries

carlosmiei committed:
fix response tests and others

carlosmiei committed:
fix response

carlosmiei committed:
try skip response tests

carlosmiei committed:
dynamically load binary in tests

carlosmiei committed:
fix ts/js loading

carlosmiei committed:
response tests passing

carlosmiei committed:
fix php sync/async version

carlosmiei committed:
missing test_helpers

carlosmiei committed:
fix c# and php

carlosmiei committed:
fix blocking method go

carlosmiei committed:
skip go tests for now

ttodua committed:
Change runner to ubuntu-24.04-arm64

ttodua committed:
Modify CI workflow for Ubuntu version and swap setup

Updated the GitHub Actions workflow to use the latest Ubuntu version and added swap file setup for testing.

ttodua committed:
revert

carlosmiei committed:
add debug statmenets

sc0Vu committed:
Merge branch 'master' into add-lighter

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into add-lighter

carlosmiei committed:
Merge branch 'add-lighter' of github.com:sc0Vu/ccxt into add-lighter

carlosmiei committed:
try to debug session

carlosmiei committed:
add debug step

carlosmiei committed:
remove set

carlosmiei committed:
update command

carlosmiei committed:
fix typo

carlosmiei committed:
rm

carlosmiei committed:
add more logs

carlosmiei committed:
readd tmate

carlosmiei committed:
disable lighter in C# for now

carlosmiei committed:
rm logs

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into add-lighter

carlosmiei committed:
rm defaults

carlosmiei committed:
skip test props

carlosmiei committed:
update docs

sc0Vu committed:
lighter: update candels

sc0Vu committed:
Merge branch 'master' into add-lighter

carlosmiei committed:
fix response tests

sc0Vu committed:
lighter: update

carlosmiei committed:
Merge branch 'add-lighter' of github.com:sc0Vu/ccxt into add-lighter

carlosmiei committed:
skip more props

carlosmiei committed:
Merge branch 'add-lighter' of github.com:sc0Vu/ccxt into add-lighter

sc0Vu committed:
lighter: update

carlosmiei committed:
rm

carlosmiei committed:
Merge branch 'add-lighter' of github.com:sc0Vu/ccxt into add-lighter

carlosmiei committed:
add dex flag
